fix xmonad/xmonad#87

Javran opened the pull request:
### Description

Fix #87. Switch focus when mouse enters a workspace but not moving into a window (w' = 0).

Attempted to preserve the old fix (https://github.com/xmonad/xmonad/commit/3a140badf52a7ce5676a3dc240265be136593f9c) while recover the intended behavior.

However I can only say for sure that this works for me:

- I tried to follow the discussion found in https://github.com/xmonad/xmonad/commit/3a140badf52a7ce5676a3dc240265be136593f9c, but that's a very old thread and many people had reported not reproducing the bug so I don't bother testing it
- attempted `xmonad-testing`, can confirm nothing beyond that there's no problem with default config, the bug occurs only when you have more than one monitor and I have no idea how could xephyr simulate an environment like that.

Also updated `CHANGES.md`, but can't come up with a more concise and clear description, so suggestions on that is also welcomed.

### Checklist

  - [x] I've read [CONTRIBUTING.md](https://github.com/xmonad/xmonad/blob/master/CONTRIBUTING.md)

  - [x] I've confirmed these changes don't belong in xmonad-contrib instead

  - [x] ~~I tested my changes with [xmonad-testing](https://github.com/xmonad/xmonad-testing)~~ (tested with default settings (`main = xmonad def`) on my own machine (see comment below))

  - [x] I updated the `CHANGES.md` file

geekosaur commented:
On Thu, Mar 9, 2017 at 5:26 PM, Javran Cheng <notifications@github.com>
wrote:

> - attempted xmonad-testing, can confirm nothing beyond that there's no
> problem with default config, the bug occurs only when you have more than
> one monitor and I have no idea how could xephyr simulate an environment
> like that.


xephyr can't; you'd need to use xnest.

-- 
brandon s allbery kf8nh                               sine nomine associates
allbery.b@gmail.com                                  ballbery@sinenomine.net
unix, openafs, kerberos, infrastructure, xmonad        http://sinenomine.net

Javran commented:
@geekosaur I take it that `xmonad-testing` is just for showing how to make it work for some custom build systems and setting up an isolated environment to try xmonad configs. Would it be good enough if I compile the config with default settings and run it in my host machine?

geekosaur commented:
That should be good enough. `xmonad-testing` is a nice idea but its underpinnings aren't really sufficient for all testing scenarios --- and it's not clear that they ever can be for anyone but X11 experts.

Javran commented:
Tested with default settings and confirmed it's working.

commands are:

```bash
$ git rev-parse HEAD
200afa4d9f853a3fff62f4107700404656838779
$ cat /tmp/config.hs
module Main where
import XMonad
main :: IO ()
main = xmonad def
$ rm ~/.xmonad/xmonad ~/.xmonad/xmonad-x86_64-linux
$ stack exec -- ghc -O2 /tmp/config.hs -main-is Main -o ~/.xmonad/xmonad
$ cp ~/.xmonad/xmonad ~/.xmonad/xmonad-x86_64-linux
$ <hit xmonad replace key>
```

pjones commented:
xephyr can indeed emulate more than one monitor.  Just supply more than
one set of dimensions.  I've been meaning to make this an option in the
xmonad-testing script.

pjones commented:
Oh, and thanks for the patch.  I'm hoping to review it soon.

dj-foxxy commented:
Be using this fix for ~2 months without issue. Please merge.

eldog commented:
Can confirm this works for me, fixes the very frustrating behaviour experienced when using multiple desktops.

pjones commented on src/XMonad/Main.hs:
```diff
@@ -368,7 +368,7 @@ handle e@(CrossingEvent {ev_window = w, ev_event_type = t})
         root <- asks theRoot
         (_, _, w', _, _, _, _, _) <- io $ queryPointer dpy root
-        when (w == w') (focus w)
+        when (w' == 0 || w == w') (focus w)
```
Is `w' == 0` testing to see if `w'` is the root window?  If so please just compare to the `root` binding and I'll merge.

Javran replied:
I also wanted to eliminate the magical `0` there, but my knowledge about xlib is limited:
`w' == root` is not the same as `w' == 0`, because turns out `root` is not `0` in this case.
What I do know is by debugging I found that when a mouse is moved to another workspace but without moving into any window, `w'` becomes `0` - I've searched some docs but still don't know what that `0` means.

I've also tried changing the binding line to `(flg, _, w', x', y', wx', wy', _) <- io $ queryPointer dpy root`, but `flg` is always `True`, none of `x'`, `y'`, `wx'` or `wy'` seems to be useful either.

pjones replied:
From @geekosaur's comment it sounds like you should change `0` to `None`.

geekosaur replied:
Not that simple, I'm afraid; I was quoting the C API, and the Haskell wrapper can't use `None` that way. The C `None` is *not* reflected through to Haskell; unless you want to add something to the X11 package, `0` is correct.

pjones replied:
Okay, @Javran it sounds like the correct solution is to just add a comment describing what `0` is.  After that I'll merge the PR.

geekosaur commented:
It's not returning the root window, it's returning `None` (which is indeed 0).

Javran committed:
fix xmonad/xmonad#87

switch focus when mouse is entering a workspace
but not moving into a window (w' == 0)

for magic number 0 in w' == 0
see discussion in https://github.com/xmonad/xmonad/pull/88#pullrequestreview-62489502

Javran commented:
Commented & conflict resolved (rebased). see if it looks good.

Edit: not sure why CI is failing, but it seemed to be a problem with X11.

pjones commented:
I don't know what's going on with the build but your change is so simple I merged anyways.  Thanks!

geekosaur commented:
For some reason the X11 package is willing to build without xinerama headers/libs but fails without XRandR headers/libs. (A situation where you have only the latter is impossible, I think. So this might be a bug in X11's configure.)

geekosaur commented:
Also arguably our CI setup should make sure both xinerama and xrandr dev libs are installed.

colonelpanic8 commented:
@pjones @geekosaur This issue with building X11 actually started appearing in the taffybar build recently.

I think it might be because travis CI recently started building using trusty instead of precise.

I was able to fix it for the time being by adding
```
sudo: required
```
and explicitly requiring the necessary dependencies,  but we should probably figure out why there is a difference in trusty and precise.
